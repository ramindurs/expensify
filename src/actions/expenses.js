import {v4 as uuidv4} from "uuid";
import database from '../firebase/firebase';

const FB_REF = 'expenses';

export const addExpense = (expense) => ({
        type: 'ADD_EXPENSE',
        expense
    }
);

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {
            description,
            note,
            amount,
            createdAt
        };
        return database().ref(FB_REF).push(expense).then((fbRef) => {
            dispatch(addExpense({
                id: fbRef.key,
                ...expense
            }))
        })
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (id) => {
    return (dispatch) => {
        const ref = `${FB_REF}/${id}`;
        return database().ref(ref).remove().then(() => {
            return dispatch(removeExpense(id));
        }).catch((error) => {
            console.error('Error removing ', ref);
        });
    };
};

export const editExpense = ({id, updates}) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = ({id, updates}) => {
    return (dispatch) => {
        const ref = `${FB_REF}/${id}`;
        return database().ref(ref).update(updates).then(() => {
            return dispatch(editExpense({id, updates}));
        }).catch((error) => {
            console.error('Error updating', ref);
        });
    }
};

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database().ref(FB_REF).once('value')
            .then((dataSnapshot) => {
                const expenses = [];
                dataSnapshot.forEach((childSnapshot) => {
                    expenses.push({
                        ...childSnapshot.val(),
                        id: childSnapshot.key
                    });
                });
                return dispatch(setExpenses(expenses));
            });
    };
};
