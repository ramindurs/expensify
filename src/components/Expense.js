import React from 'react';
import {Link} from 'react-router-dom';
import moment from "moment";
import numeral from 'numeral';

export const Expense = ({ id, description, amount, createdAt, dispatch, history}) => {
    return (
        <div>
            {id} <Link to={`/edit/${id}` }>{description}</Link>
            <p>
                {numeral(amount / 100).format('0,0.00')}
                -
                {moment(createdAt).format('MMM Do YYYY')}
            </p>
        </div>
    );
};

export default Expense;
