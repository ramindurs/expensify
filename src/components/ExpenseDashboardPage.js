import React from 'react';
import Expenses from "./Expenses";
import ExpensesFilter from "./ExpensesFilter";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesFilter/>
        <ExpensesSummary/>
        <Expenses/>
    </div>
);

export default ExpenseDashboardPage;