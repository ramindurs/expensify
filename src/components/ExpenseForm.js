import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from "react-dates";


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        const expense = props.expenseToEdit;
        this.state = {
            description: expense && expense.description ? expense.description : '',
            note: expense && expense.note ? expense.note : '',
            amount: expense && expense.amount ? this.convertIntToFloatString(expense.amount) : '',
            createdAt: expense && expense.createdAt ? moment(expense.createdAt) : moment(),
            calendarFocus: false,
            error: ''
        };
    }
    convertIntToFloatString = (intnum) => parseFloat(intnum) / 100;

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (date) => {
        if (!!date) {
            this.setState(() => ({createdAt: date}));
        }
    };
    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocus: focused}));
    };
    onSubmitForm = (e) => {
        e.preventDefault();
        if (this.state.description && this.state.amount) {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: Math.round(parseFloat(this.state.amount, 10) * 100),
                createdAt: this.state.createdAt.valueOf()
            });
        } else {
            this.setState(() => ({error: 'Please provide description and amount'}));
        }
    };

    render() {
        return (
            <div>
                {this.state.error ? <p>{this.state.error}</p> : ''}
                <form onSubmit={this.onSubmitForm}>
                    <input
                        type="text"
                        placeholder={'description'}
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoFocus/>
                    <input
                        type="text"
                        placeholder={'amount'}
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocus}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        onChange={this.onNoteChange}
                        value={this.state.note}
                        placeholder={'Add a note'}>
                    </textarea>
                    <button>{this.props.submitMessage}</button>
                </form>
            </div>
        )
    };
};
