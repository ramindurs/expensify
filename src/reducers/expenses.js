
const expensesRedecuerDefaultState = [];

export default (state = expensesRedecuerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(ex => ex.id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id !== action.id ? expense : {
                ...expense,
                ...action.updates
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};