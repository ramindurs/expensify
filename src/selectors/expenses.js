import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true;
        const textMatch = expense.description.toUpperCase().indexOf(text.toUpperCase()) !== -1;
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => (sortBy === 'date' ? b.createdAt - a.createdAt: b.amount - a.amount));
};