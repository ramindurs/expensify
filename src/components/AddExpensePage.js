import React from 'react';
import {connect} from 'react-redux';
import {startAddExpense} from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <div className='page-header-layout'>
                    <div className="content_container">
                        <h1 className='page-header-layout__title'>Add Expenses</h1>
                    </div>
                </div>
                <div className="content_container">
                    <ExpenseForm
                        submitMessage={'Save Expense'}
                        onSubmit={this.onSubmit}/>
                </div>
            </div>
        );
    }
}

const mapDispatchTpProps = (dispatch) => {
    return {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDispatchTpProps)(AddExpensePage);
