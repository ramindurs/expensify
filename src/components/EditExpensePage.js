import React from 'react';
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component {
    removeExpense = (id) => {
        this.props.startRemoveExpense(id);
        this.props.history.push('/dashboard');
    };
    editExpense = (id, expense) => {
        this.props.startEditExpense(id, expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <h3>Edit Expense ID {this.props.expense.id}</h3>
                <ExpenseForm
                    expenseToEdit={this.props.expense}
                    submitMessage={'Update Expense'}
                    onSubmit={(expense) => {
                        this.editExpense(this.props.expense.id, expense)
                    }}
                />
                <button onClick={() => {
                    this.removeExpense(this.props.expense.id)
                }
                }>Delete
                </button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(e => e.id === props.match.params.id)
    };
};

const matchDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense({
            id: id,
            updates: expense
        })),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    };
};

export default connect(mapStateToProps, matchDispatchToProps)(EditExpensePage);
