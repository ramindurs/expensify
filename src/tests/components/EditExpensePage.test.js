import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensePage} from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let wrapper, expense, history, removeExpenseSpy, editExpenseSpy;

beforeEach(() => {
    expense = expenses[0];
    history = {
        push: jest.fn()
    };
    removeExpenseSpy = jest.fn();
    editExpenseSpy = jest.fn();

    wrapper = shallow(<EditExpensePage
        expense={expense}
        history={history}
        startRemoveExpense={removeExpenseSpy}
        editExpense={editExpenseSpy}
    />);
});

test('should render edit expense page', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click');

    expect(removeExpenseSpy).toHaveBeenLastCalledWith('1');
});

test('should handle update expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense);

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpenseSpy).toHaveBeenLastCalledWith('1', expense);
});