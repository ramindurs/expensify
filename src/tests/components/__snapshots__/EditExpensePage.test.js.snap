// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page 1`] = `
<div>
  <h3>
    Edit Expense ID 
    1
  </h3>
  <ExpenseForm
    expenseToEdit={
      Object {
        "amount": 1500,
        "createdAt": 1000,
        "description": "Water Bill",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
    submitMessage="Update Expense"
  />
  <button
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
